### PR TITLE
Core Data: Replace unnecessary fetch requests for upserting orders and order search results

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #14090 

## Description

This PR follows the suggestion in p91TBi-cl2-p2 to clean up unnecessary and expensive fetch requests made in for loops for orders including:
- Order search results
- Order items and item taxes
- Order coupons
- Order fees
- Order shipping lines & shipping line taxes
- Order taxes

## Steps to reproduce

- Log in to a test store with existing products.
- Create an order by selecting one or more products, and adding coupons, fees, shipping, and taxes. Save the changes.
- Open the new order from the order list and confirm that all details are correct.
- Tap the Search button and search for any order. Confirm that the feature works as expected.

## Testing information

Tested with simulator iPhone 15 Pro iOS 17.4 and confirmed that order deails and order search work correctly.
All unit tests still pass.

## Screenshots

N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

TODO: @itsmeichigo to update the release notes before merging.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Storage/Storage/Tools/StorageType+Extensions.swift [hunk 1]
```diff
@@ -35,6 +35,14 @@ public extension StorageType {
 
     // MARK: - Orders
 
+    /// Retrieves the Stored Orders given the IDs.
+    ///
+    func loadOrders(siteID: Int64, orderIDs: [Int64]) -> [Order] {
+        let predicate = NSPredicate(format: "orderID in %@", orderIDs)
```
Comment (selanthiraiyan) [inline Storage/Storage/Tools/StorageType+Extensions.swift]: Do we want to add the `siteID` to the predicate? With the current predicate, I am afraid that we will end up loading orders from other sites with the same order ID.

File: Storage/StorageTests/Tools/StorageTypeExtensionsTests.swift [hunk 2]
```diff
@@ -66,6 +66,30 @@ final class StorageTypeExtensionsTests: XCTestCase {
         XCTAssertEqual(site, storedSite)
     }
 
+    func test_loadOrders_list_by_siteID_and_orderIDs() {
```
Comment (selanthiraiyan) [inline Storage/StorageTests/Tools/StorageTypeExtensionsTests.swift:69]: Suggestion: We could update this test by adding an order with a different site ID other than `sampleSiteID` to validate that only orders from the given `siteID` are loaded.

File: Storage/Storage/Tools/StorageType+Extensions.swift (hunk 1 above)
Comment (itsmeichigo) [reply Storage/Storage/Tools/StorageType+Extensions.swift]: Good catch! I updated the predicate in a27533c7133933d91e0cf80da4a3e24259c73ddd.

File: Storage/StorageTests/Tools/StorageTypeExtensionsTests.swift (hunk 2 above)
Comment (itsmeichigo) [reply Storage/StorageTests/Tools/StorageTypeExtensionsTests.swift:69]: Updated the test in a27533c7133933d91e0cf80da4a3e24259c73ddd.

**Comment by itsmeichigo** [comment]
Thanks @selanthiraiyan for the reviews! I addressed your comment with a commit above, this PR is ready for another look.

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift [hunk 3]
```diff
@@ -256,11 +253,10 @@ struct OrdersUpsertUseCase {
     /// Updates, inserts, or prunes the provided StorageShippingLine's taxes using the provided read-only ShippingLine
     ///
     private func handleShippingLineTaxes(_ readOnlyItem: Networking.ShippingLine, _ storageItem: Storage.ShippingLine, _ storage: StorageType) {
-        let shippingID = readOnlyItem.shippingID
 
         // Upsert the taxes from the read-only orderItem
         for readOnlyTax in readOnlyItem.taxes {
-            if let existingStorageTax = storage.loadShippingLineTax(shippingID: shippingID, taxID: readOnlyTax.taxID) {
+            if let existingStorageTax = storageItem.taxes?.first(where: { $0.taxID == readOnlyTax.taxID }) {
```
Comment (selanthiraiyan) [inline Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:259]: I see that `loadShippingLineTax` method checks for both `shippingID` and `taxID`. Do we have to add the additional check of `shippingID` here?

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift [hunk 4]
```diff
@@ -137,11 +135,10 @@ struct OrdersUpsertUseCase {
     /// Updates, inserts, or prunes the provided StorageOrderItem's taxes using the provided read-only OrderItem
     ///
     private func handleOrderItemTaxes(_ readOnlyItem: Networking.OrderItem, _ storageItem: Storage.OrderItem, _ storage: StorageType) {
-        let itemID = readOnlyItem.itemID
 
         // Upsert the taxes from the read-only orderItem
         for readOnlyTax in readOnlyItem.taxes {
-            if let existingStorageTax = storage.loadOrderItemTax(itemID: itemID, taxID: readOnlyTax.taxID) {
+            if let existingStorageTax = storageItem.taxes?.first(where: { $0.taxID == readOnlyTax.taxID }) {
```
Comment (selanthiraiyan) [inline Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:141]: Same as the other comment. 

I see that `loadOrderItemTax` method checks for both `itemID` and `taxID`. Do we have to add the additional check of `itemID` here?

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift (hunk 3 above)
Comment (itsmeichigo) [reply Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:259]: The `storageItem` is sent from `handleOrderShippingLines`, which is the stored shipping line filtered by `shippingID`. That's why checking for `shippingID` here would be redundant.

File: Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift (hunk 4 above)
Comment (itsmeichigo) [reply Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:141]: Similar as above, `storageItem` here is the order item filtered by `itemID` in `handleOrderItems`. Filtering by `itemID` would be redundant.

**Comment by itsmeichigo** [comment]
@selanthiraiyan I responded to your questions, please let me know if anything is still unclear. Thanks!

**Review by selanthiraiyan — APPROVED** [review]
Thanks for the updates and clearing my questions, Huong! 

The code LGTM and works as described. 🚀

**Pull request merged**